File: MiniDOMBase/DMD/Anagrafica/Tasks/Handlers/BaseTaskHandler.cs

namespace minidom
{
    public partial class Anagrafica
    {

        /// <summary>
        /// Classe base che implementa l'interfaccia <see cref="ITaskHandler"/>
        /// </summary>
        public abstract class BaseTaskHandler 
            : ITaskHandler
        {
            private TaskLavorazione m_Result = null;

            /// <summary>
            /// Costruttore
            /// </summary>
            public BaseTaskHandler()
            {
                ////DMDObject.IncreaseCounter(this);
                m_Result = null;
            }

            /// <summary>
            /// Risultati restituiti dell'handler
            /// </summary>
            public TaskLavorazione Result
            {
                get
                {
                    return m_Result;
                }
            }

            /// <summary>
            /// Gestisce la regola
            /// </summary>
            /// <param name="taskSorgente"></param>
            /// <param name="regola"></param>
            /// <param name="parametri"></param>
            /// <returns></returns>
            public TaskLavorazione Handle(TaskLavorazione taskSorgente, RegolaTaskLavorazione regola, string parametri)
            {
                m_Result = new TaskLavorazione();
                m_Result.TaskSorgente = taskSorgente;
                m_Result.RegolaEseguita = regola;
                m_Result.ParametriAzione = parametri;
                m_Result.PuntoOperativo = taskSorgente.PuntoOperativo;
                m_Result.DataInizioEsecuzione = DMD.DateUtils.Now();
                m_Result.Stato = ObjectStatus.OBJECT_VALID;
                m_Result.RisultatoAzione = "";
                m_Result.Save();
                // Me.m_Result.RisultatoAzione = Me.HandleInternal()
                HandleInternal();
                if (taskSorgente is object)
                {
                    taskSorgente.TaskDestinazione = m_Result;
                    taskSorgente.Save();
                }

                m_Result.Save();
                return m_Result;
            }

            /// <summary>
            /// Gestisce la regola internamente
            /// </summary>
            protected abstract void HandleInternal();

            /// <summary>
            /// Distruttore
            /// </summary>
            ~BaseTaskHandler()
            {
                ////DMDObject.DecreaseCounter(this);
            }
        }
    }
}

File: MiniDOMBase/DMD/Anagrafica/Tasks/BaseTaskHandler.vb
Imports minidom
Imports minidom.Sistema
Imports minidom.Databases



Partial Public Class Anagrafica
     
    Public MustInherit Class BaseTaskHandler
        Implements ITaskHandler

        Private m_Result As TaskLavorazione = Nothing

        Public Sub New()
            DMDObject.IncreaseCounter(Me)
            Me.m_Result = Nothing
        End Sub

        Public ReadOnly Property Result As TaskLavorazione
            Get
                Return Me.m_Result
            End Get
        End Property
         

        Public Function Handle(ByVal taskSorgente As TaskLavorazione, ByVal regola As RegolaTaskLavorazione, ByVal parametri As String) As TaskLavorazione Implements ITaskHandler.Handle
            Me.m_Result = New TaskLavorazione
            Me.m_Result.TaskSorgente = taskSorgente
            Me.m_Result.RegolaEseguita = regola
            Me.m_Result.ParametriAzione = parametri
            Me.m_Result.PuntoOperativo = taskSorgente.PuntoOperativo
            Me.m_Result.DataInizioEsecuzione = DateUtils.Now
            Me.m_Result.Stato = ObjectStatus.OBJECT_VALID
            Me.m_Result.RisultatoAzione = ""
            Me.m_Result.Save()
            'Me.m_Result.RisultatoAzione = Me.HandleInternal()
            Me.HandleInternal()
            If (taskSorgente IsNot Nothing) Then
                taskSorgente.TaskDestinazione = Me.m_Result
                taskSorgente.Save()
            End If
            Me.m_Result.Save()
            Return Me.m_Result
        End Function

        Protected MustOverride Sub HandleInternal()

        Protected Overrides Sub Finalize()
            MyBase.Finalize()
            DMDObject.DecreaseCounter(Me)
        End Sub
    End Class


End Class